API/NoDocs/Microsoft.Extensions.Http.Resilience/EndpointGroup.cs

```diff
@@ -0,0 +1,16 @@
+// Assembly 'Microsoft.Extensions.Http.Resilience'
+
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Microsoft.Extensions.Options;
+using Microsoft.Shared.Data.Validation;
+
+namespace Microsoft.Extensions.Http.Resilience;
+
+public class EndpointGroup
+{
+    [Microsoft.Shared.Data.Validation.Length(1)]
+    [ValidateEnumeratedItems]
+    public IList<WeightedEndpoint> Endpoints { get; set; }
```

Comment:
@bartonjs what's our general view on this

Reply:
* DO NOT use ArrayList or `List<T>` in public APIs.  (I'm sure we use `List<T>` all over the place; it says to prefer `Collection<T>` or the interfaces, but I'm pretty sure later it'll say don't use the interfaces for properties)
* DO NOT provide settable collection properties. (remove the setter, they should call Clear/Add/AddRange)

I'm not quickly finding the "don't use interfaces for property returns" guideline written down, but our general mantra is (to quote Anthony Moore in FDG pseudo-quoting the design meetings) "require the weakest thing you need, and return the strongest thing you have.".  If the ctor initializes that property to `new List<WeightedEndpoint>()`, there's someone somewhere who will end up hard-casting it for some reason (or even as-casting and not implementing the fallback correctly), and in the BCL we'd say that changing to some other type later is a breaking change.  At that point we say to just own up and declare it as `List<WeightedItem>` and call it a day. If you're using a custom IList, and the type is non-public, then `IList<WeightedItem>` is the correct type as there's nothing more specific a caller can depend on.

But, basically, the design guidelines say it should be one of

```C#
// If it's really just List<T>
public List<WeightedEndpoint> Endpoints { get; }
```

```C#
// If it's some other public IList type
public SomeOtherList<WeightedEndpoint> Endpoints { get; }
```

```C#
// The actual type is opaque, but you're still not supposed to have a setter.
public IList<WeightedEndpoint> Endpoints { get; }
```

There are extenuating circumstances, such as this being a pure input object and there being important reasons why someone should be able to provide some other kind of IList, which would "allow" the setter to remain.

Reply:
We use this pattern all over the place in this repo, primarily in options POCO. The caller can supply whatever type of list suits them, and it will be ingested by the implementation. In other words, this is not about returning state to someone, it's about consuming it. This is not something that is intended to change dynamically to reflect some live state, it's just info being passed by the caller into an implementation.

We have setters so as not to constraint the caller that may have this state coming from different places/shapes.

Reply:
In this case, should it be declared as nullable?